File: twoSum.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // prevMap stores <Value, Index>
        Map<Integer, Integer> prevMap = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int diff = target - num;
            
            // Check if the complement exists in the map
            if (prevMap.containsKey(diff)) {
                return new int[] { prevMap.get(diff), i };
            }
            
            // If not found, add current number and its index to map
            prevMap.put(num, i);
        }
        
        // Return empty array if no solution is found
        return new int[] {};
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        
        int[] result = sol.twoSum(nums, target);
        
        if (result.length == 2) {
            System.out.println("[" + result[0] + ", " + result[1] + "]");
        }
    }
}

File: twoSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // prevMap stores {Value: Index}
        unordered_map<int, int> prevMap;
        
        for (int i = 0; i < nums.size(); i++) {
            int diff = target - nums[i];
            
            // Check if the complement exists in the map
            // .find() returns .end() if the key is not found
            if (prevMap.find(diff) != prevMap.end()) {
                return {prevMap[diff], i};
            }
            
            // If not found, add current number and index to map
            prevMap[nums[i]] = i;
        }
        
        // Return empty vector if no solution is found
        return {};
    }
};

int main() {
    Solution sol;
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    
    vector<int> result = sol.twoSum(nums, target);
    
    if (!result.empty()) {
        cout << "[" << result[0] << ", " << result[1] << "]" << endl;
    }
    
    return 0;
}
